src/EditorFeatures/Core/Workspaces/ITextBufferVisibilityTracker.cs

```diff
@@ -37,40 +39,59 @@ internal interface ITextBufferVisibilityTracker
     internal static class ITextBufferVisibilityTrackerExtensions
     {
         /// <summary>
-        /// Waits the specified amount of time while the specified <paramref name="subjectBuffer"/> is not visible.  If any
-        /// document visibility changes happen, the delay will cancel.
+        /// Waits the specified amount of time while the specified <paramref name="subjectBuffer"/> is not visible.  If
+        /// any document visibility changes happen, the delay will cancel.
         /// </summary>
-        public static async Task DelayWhileNonVisibleAsync(
+        public static Task DelayWhileNonVisibleAsync(
             this ITextBufferVisibilityTracker? service,
             IThreadingContext threadingContext,
+            IAsynchronousOperationListener listener,
             ITextBuffer subjectBuffer,
             TimeSpan timeSpan,
             CancellationToken cancellationToken)
         {
             // Only add a delay if we have access to a service that will tell us when the buffer become visible or not.
             if (service is null)
-                return;
+                return Task.CompletedTask;
 
-            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-            if (service.IsVisible(subjectBuffer))
-                return;
+            // Because cancellation is both expensive, and a super common thing to occur while we're delaying the caller
+            // until visibility, we special case the implementation here and transition to the canceled state
+            // explicitly, rather than throwing a cancellation exception.
 
-            // ensure we listen for visibility changes before checking.  That way we don't have a race where we check
-            // something see it is not visible, but then do not hear about its visibility change because we've hooked up
-            // our event after that happens.
-            var visibilityChangedTaskSource = new TaskCompletionSource<bool>();
-            var callback = void () => visibilityChangedTaskSource.TrySetResult(true);
-            service.RegisterForVisibilityChanges(subjectBuffer, callback);
+            var taskOfTask = DelayWhileNonVisibleWorkerAsync().ContinueWith(
+                task => cancellationToken.IsCancellationRequested ? Task.FromCanceled(cancellationToken) : task,
```

Comment:
Could be updated to avoid hiding failures?

```suggestion
                task => cancellationToken.IsCancellationRequested && task.Status == TaskStatus.RanToCompletion ? Task.FromCanceled(cancellationToken) : task,
```

Reply:
done.